Add spin and singlet operators to HubbardOperators

Yue-Zhengyuan opened the pull request:
This PR adds spin and singlet operators to HubbardOperators.

- The spin operators are defined as $S^a_i = c^\dagger_{i\alpha} \sigma^a_{\alpha \beta} c_{i\beta} / 2$ (a = x, y, z). Their nonzero matrix elements are the same as their counterpart in t-J model (they don't have nonzero elements relevant to the doubly-occupied state). 
- The singlet operator is still defined as $\epsilon_{\alpha \beta} c_{i\alpha} c_{j\beta}$. It actually have SU(2) spin symmetry, which remains to be added for both t-J and Hubbard models.

In addition, the definition of `u_min_u_plus`, `d_min_d_plus` in `HubbardOperators` are changed to be consistent with #15, which was overlooked there. This will then close issue #17.

TODO:

- [x] Add test to ensure that two-body `S_plus_S_min` (and its h.c.) and `S_exchange` with non-trivial spin symmetry is correctly defined now that the physical space has the doubly-occupied state.

sanderdemeyer commented:
Overall this looks good to me.

Apart from the test on `S_plus_S_min` and `S_exchange`, I don't know whether the current tests are strict enough on the operators `d_min_u_min` and `u_min_d_min`. I have checked this theoretically and I come to the same conclusion as you, but if you e.g. change the minus sign of the last fusion tree of the two operators, the tests still pass. 

Is there a reason you are only interested in these operators, and not in e.g. `d_plus_u_plus` or `d_plus_u_min`? If you were planning to add them anyway in the near future, maybe we can make an effort to do this now (and also use them to add some additional tests).

Yue-Zhengyuan commented:
I'm interested in the pattern of singlet pairing in t-J/Hubbard states with finite hole doping, 
```math
(\Delta_{sin})_{ij} = \epsilon_{\alpha \beta} c_{i\alpha} c_{j\beta}
```
which give us properties of superconductivity. Actually, `d_plus_u_plus` (or its h.c. `u_min_d_min`) is also useful, since it is related to the triplet pairing
```math
(\Delta_{tri})^a_{ij} = \frac{1}{\sqrt{2}} c_{i\alpha} (i \sigma^y \sigma^a)_{\alpha \beta} c_{j\beta}
\quad (a = x, y, z)
```
but it's less dominant in t-J/Hubbard model. The three components of $\Delta_{tri}$ are proportional to

- x:  `u_min_u_min - d_min_d_min`
- y:  `u_min_u_min + d_min_d_min`
- z:  `u_min_d_min + d_min_u_min`

For `d_plus_u_min` etc, I haven't ever seen a physical model in which hopping will also change the spin. So I think it's OK we don't add these operators. 

@lkdvos What do you suggest about the tests?

sanderdemeyer commented:
Maybe we can just add `d_plus_u_plus` and `u_plus_d_plus` and check whether `d_plus_u_plus' = -d_min_u_min` and `u_plus_d_plus' = -u_min_d_min`?

Yue-Zhengyuan commented:
I want to *define* `d_plus_u_plus = -adjoint(d_min_u_min)` etc to save manual assignment of matrix elements; this will then need some other tests...

lkdvos commented on src/hubbardoperators.jl:
````diff
@@ -3,19 +3,27 @@ module HubbardOperators
-Return the local hilbert space for a Hubbard-type model with the given particle and spin symmetries.
+Return the local hilbert space for a Hubbard-type model with the given particle and spin symmetries. The four basis states are
+```
+    |0⟩ (vacuum), |↑⟩ = (c↑)†|0⟩, |↓⟩ = (c↓)†|0⟩, |2⟩ = (c↑)†(c↓)†|0⟩.
````
```suggestion
    |0⟩ (vacuum), |↑⟩ = (c↑)†|0⟩, |↓⟩ = (c↓)†|0⟩, |↑↓⟩ = (c↑)†(c↓)†|0⟩.
```
I like this slightly more since it explicitly adds the order of the doubly occupied state, but I'm okay with both

Yue-Zhengyuan replied:
I chose |2⟩ because I thought |↑↓⟩ might be confused with |↑_1, ↓_2⟩. So I prefer |2⟩ as long as we make it clear about the internal order.

lkdvos replied:
I guess for two sites we add a comma, but I see your point. In general we haven’t been very consistent with the Unicode anyways I think. I also just discovered there is ⇅ as well.

Yue-Zhengyuan replied:
Now that I'm in the minority, I just changed the notation to `↑↓`. Different sites are separated by commas.

lkdvos commented on src/hubbardoperators.jl:
```diff
@@ -214,6 +222,94 @@ function ud_num(elt::Type{<:Number}, ::Type{U1Irrep}, ::Type{SU2Irrep})
 end
 const nꜛꜜ = ud_num
 
+@doc """
```
For these spin operators, could you also add an explicit error that they require trivial symmetry?

Yue-Zhengyuan replied:
I'll add ArgumentError for `S_plus`. Since `S_min`, `S_x`, `S_y` call `S_plus`, they will also throw this ArgumentError. I want to do the same for the t-J spin operators as well, since they are closely related (thought ideally it should be in a separate PR).

Yue-Zhengyuan commented:
@sanderdemeyer I added (quite) a few things, so this PR now also involves many changes not quite related to its title; sorry! 

- I discover that in `HubbardOperators`, although `e_min_e_plus` has changed to the new convention, `u_min_u_plus` and `d_min_d_plus` were not. So I changed them for consistency. 
- I improved test coverage on `S_exchange` by checking its eigenvalues from the trivial and the symmetric versions. 
- Some tests on the hopping terms that are just repetitions of their definition are removed.

I suggest we postpone addition of the triplet terms to a separate PR.

sanderdemeyer commented:
Apart from the minor remark about the notation for the doubly occupied state, everything looks good to me. Indeed postponing the addition of operators like `d_plus_u_plus` and the triplet state to another PR seems like the best idea.

sanderdemeyer commented:
Correct me if I'm wrong, but I think the `singlet_min` operator is not symmetric but antisymmetric with respect to SU(2). I can add a SU(2) symmetric version of the triplet operator (in the next PR) though.

Yue-Zhengyuan commented:
@sanderdemeyer How about we merge this PR for now and leave the triplet pairing and SU(2) version of singlet pairing to a future PR? In addition, `e_hopping` is also SU(2) invariant, and we can add it later together with the singlet pairing.
